Add project structure chapter (#14337)
dotnet/docs: docs/architecture/blazor-for-web-forms-developers/project-structure.md

---
title: Project structure for Blazor apps
description: Comparing the project structure of ASP.NET Web Forms and Blazor projects.
author: danroth27
ms.author: daroth
ms.date: 09/11/2019
---

# Project structure for Blazor apps

Blazor apps share a lot in common with ASP.NET Web Froms apps. Even though the projects might look quite different, the concepts have many similarities. Here we will look at the structure of a Blazor project and compare it to ASP.NET Web Forms projects.

> To create your first Blazor app follow the instructions in the [Blazor getting started steps](https://docs.microsoft.com/aspnet/core/blazor/get-started). You can follow the instructions to create either a Blazor Server app or a Blazor WebAssembly app hosted in ASP.NET Core. Most of the code in both projects is the same, although the hosting model specific logic will differ.

Reviewer: I hope you don't mind that I've added a comma to your suggestion @scottaddie

> To create your first Blazor app follow the instructions in the [Blazor getting started steps](https://docs.microsoft.com/aspnet/core/blazor/get-started). You can follow the instructions to create either a Blazor Server app or a Blazor WebAssembly app hosted in ASP.NET Core. Most of the code in both projects is the same, although the hosting model specific logic will differ.

Reviewer 2: Don't mind at all. Always nice to have a 2nd pair of eyes.

## Project file

Blazor Server apps are .NET Core projects. The project file for the Blazor Server app is about as simple as it can get:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
  </PropertyGroup>

</Project>
```

The project file for a Blazor WebAssembly looks slightly more involved (exact version numbers may vary):

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>

>     <TargetFramework>netstandard2.0</TargetFramework>

Author: Moving to .NET Standard 2.1 is the plan. We're currently blocked on an IL linker issue: https://github.com/mono/mono/issues/16765. In the .NET 5 frame we may also define a browser specific target framework moniker.

>     <OutputType>Exe</OutputType>

Reviewer 3: This looks weird - it's an `exe` type but targeting `netstandard`? You explain the `netstandard` below, but maybe comment about the `exe` since usually `netstandard` is recommended for libs

Author: Yeah, I'm not actually sure why we set the OutputType to Exe. Things seem to work without it. I'll find out. The tricky thing here is that the Blazor WebAssembly details are likely to change over the upcoming months we shift focus to productizing it.

Reviewer 3: Makes sense.

Blazor apps have a `Startup` class instead of a `Global.asax` file to define the startup logic for the app. The `Startup` class is used to configure the app and any app specific services. In the Blazor Server app the `Startup` class is used to setup the endpoint for the real-time connection used by Blazor between the client browsers and the server. In the Blazor WebAssembly app the `Startup` class defines the root components for teh app and where they should be rendered. We'll take a deeper look at the `Startup` class in the [App startup](./app-startup) section.

## Static files

Unlike ASP.NET Web Forms projects, not all of the files in the Blazor project are available to be requested as static files. Only the files in the *wwwroot* folder are web addressable. This folder is referred to the app's "web root". Anything outside of the app's web root is *not* web addressable. This setup provides an additional level of security that prevents accidentally exposing project files over the web.

## Configuration

> Configuration in ASP.NET Web Forms apps is typically handled using one or more *web.config* files. Blazor apps don't typically have *web.config* files. If they do, the file is only used to configure IIS specific settings when hosted on IIS. Instead, Blazor Server apps use the ASP.NET Core configuration abstractions (Blazor WebAssembly apps don't currently support the same configuration abstractions, but that may be a feature added in the future). For example, the default Blazor Server app stores some settings in *appsettings.json* .

Reviewer: Included the removal of an extra space in your suggestion @scottaddie

> Configuration in ASP.NET Web Forms apps is typically handled using one or more *web.config* files. Blazor apps don't typically have *web.config* files. If they do, the file is only used to configure IIS specific settings when hosted on IIS. Instead, Blazor Server apps use the ASP.NET Core configuration abstractions (Blazor WebAssembly apps don't currently support the same configuration abstractions, but that may be a feature added in the future). For example, the default Blazor Server app stores some settings in *appsettings.json* .

Author: Yup, good point. We should cover that in the Configuration chapter. This chapter is all about the files in the projects.

```json
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*"
}
```

> We'll learn more about configuration in ASP.NET Core projects in the [Configuration](./configuration) section. .

Reviewer 3: nit: extra period

## Razor components

> Most of the rest of the files in the Blazor projects are *.razor* files. Razor is a templating language based on HTML and C# that is used to dynamically generate web UI. The *.razor* files define components that make up the UI of the app, and for the most part they are exactly the same for both the Blazor Server and Blazor WebAssembly apps. Components in Blazor are analogous to UI controls in ASP.NET Web Forms.

Reviewer 3: Most of the rest of the files -> Majority of the files

```razor
@page "/counter"
```

Routing in Blazor is handled client-side, not on the server. As the user navigates in the browser, Blazor intercepts the navigation and then renders the component with the matching route. 

The component routes are not current inferred by the component's file location like they are with *.aspx* pages (although this feature may get added in the future). Each route must be specified explicitly on the component. Putting the routable components in a *Pages* folder has no special meaning and is purely a convention.

We'll look in greater detail at routing in Blazor in the [Pages and routing](./pages-routing.md) section.

## Layout

In APS.NET Web Forms apps common page layout is handled using master pages. In Blazor apps, page layout is handled using layout components (*Shared/MainLayout.razor*). Layout components will be discussed in more detail in [Page layout](./page-layout) section.

## Bootstrapping Blazor

To bootstrap Blazor, the apps must specify where on the page the root component (*App.Razor*) should be rendered and add the corresponding Blazor framework script.

> In the Blazor Server app, the host page for root component is defined in the *_Host.cshtml*. This file defines a Razor Page, not a component. Razor Pages use Razor syntax to define a server addressable page, very much like a *.aspx* page. The `Html.RenderComponent<TComponent>(RenderMode)` method is used to define where a root level component should be rendered. The different `RenderMode` options indicate whether the component should be rendered as static content (`RenderModel.Static`), rendered interactively once a connection with the browser has been established (`RenderMode.Server`), or first prerendered and then rendered interactively (`RenderMode.ServerPrerendered`). The script reference to *_framework/blazor.server.js* establishes the real-time connection with the server and then deals with all user interactions and UI updates.

Reviewer 2: ```suggestion
In the Blazor Server app, the root component's host page is defined in the *_Host.cshtml* file. This file defines a Razor Page, not a component. Razor Pages use Razor syntax to define a server-addressable page, very much like a *.aspx* page. The `Html.RenderComponent<TComponent>(RenderMode)` method is used to define where a root-level component should be rendered. The different `RenderMode` options indicate whether the component should be rendered as static content (`RenderModel.Static`), rendered interactively once a connection with the browser has been established (`RenderMode.Server`), or first prerendered and then rendered interactively (`RenderMode.ServerPrerendered`). The script reference to *_framework/blazor.server.js* establishes the real-time connection with the server and then deals with all user interactions and UI updates.
```

Reviewer: should it be an aspx page?

    </app>

    <script src="_framework/blazor.server.js"></script>
</body>
</html>
```

In the Blazor WebAssembly app, the host page is a simple static HTML file under *wwwroot/index.html*. The `<app>` element is used to indicate where the root component should be rendered.

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title>BlazorApp2</title>
    <base href="/" />
    <link href="css/bootstrap/bootstrap.min.css" rel="stylesheet" />
    <link href="css/site.css" rel="stylesheet" />
</head>
<body>
    <app>Loading...</app>

    <script src="_framework/blazor.webassembly.js"></script>
</body>
</html>
```

The specific component to render is configured in the app's `Startup.Configure` method with a corresponding CSS selector for where the component should be rendered.

```csharp
    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
        }

        public void Configure(IComponentsApplicationBuilder app)
        {
            app.AddComponent<App>("app");
        }
    }
```

## Build output

> When the Blazor projects are built, all of the Razor component files are compiled along with the project code into a single assembly. Unlike ASP.NET Web Forms projects, Blazor does not support runtime compilation of the UI logic.

Reviewer: ```suggestion
When the Blazor projects are built, all of the Razor component files are compiled along with the project code into a single assembly. Unlike ASP.NET Web Forms projects, Blazor doesn't support runtime compilation of the UI logic.
```

## Running the app

> To run the Blazor Server app simply hit F5 in Visual Studio. Because Blazor apps do not support runtime compilation, to see the results of code changes, including component markup changes, you typically need to rebuild and restart the app when running under the the debugger. If you don't run under the debugger (Ctrl-F5) Visual Studio will watch for file changes for you and restart the app as changes are made so you can simply refresh the browser as changes are made.

Reviewer: removed an extra the in your suggestion

> To run the Blazor WebAssembly app you can either run the client project directly using the dev server, or run the server project when hosting the app with ASP.NET Core. Blazor WebAssembly apps do not support debugging using Visual Studio, so to run the app use Ctrl-F5 instead of F5. You can instead debug Blazor WebAssembly apps directly in the browser. See [Debug ASP.NET Core Blazor](https://docs.microsoft.com/aspnet/core/blazor/debug) for details.

Reviewer 2: ```suggestion
To run the Blazor WebAssembly app, either run the client project directly using the development server or run the server project when hosting the app with ASP.NET Core. Blazor WebAssembly apps do not support debugging using Visual Studio. To run the app, use `Ctrl+F5` instead of `F5`. You can instead debug Blazor WebAssembly apps directly in the browser. See [Debug ASP.NET Core Blazor](https://docs.microsoft.com/aspnet/core/blazor/debug) for details.
```

---
title: Project structure for Blazor apps

> description: Comparing the project structure of ASP.NET Web Forms and Blazor projects.

Reviewer 2: ```suggestion
description: Compare the project structures of ASP.NET Web Forms and Blazor projects.
```

author: danroth27
ms.author: daroth
ms.date: 09/11/2019
---

# Project structure for Blazor apps

Blazor apps share a lot in common with ASP.NET Web Forms apps. Though the projects look quite different, the concepts have many similarities. Here we will look at the structure of a Blazor project and compare it to an ASP.NET Web Forms project.

To create your first Blazor app, follow the instructions in the [Blazor getting started steps](https://docs.microsoft.com/aspnet/core/blazor/get-started). You can follow the instructions to create either a Blazor Server app or a Blazor WebAssembly app hosted in ASP.NET Core. Most of the code in both projects is the same. Although, the hosting model-specific logic will differ.

## Project file

Blazor Server apps are .NET Core projects. The project file for the Blazor Server app is about as simple as it can get:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
  </PropertyGroup>

</Project>
```

> The project file for a Blazor WebAssembly looks slightly more involved (exact version numbers may vary):

Reviewer 2: ```suggestion
The project file for a Blazor WebAssembly app looks slightly more involved (exact version numbers may vary):
```

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <OutputType>Exe</OutputType>
    <LangVersion>7.3</LangVersion>
    <RazorLangVersion>3.0</RazorLangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Blazor" Version="3.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.Build" Version="3.1.0" PrivateAssets="all" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.HttpClient" Version="3.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.DevServer" Version="3.1.0" PrivateAssets="all" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\Shared\BlazorWebAssemblyApp1.Shared.csproj" />
  </ItemGroup>

</Project>
```

> Blazor WebAssembly projects target .NET Standard instead of .NET Core because they run in the browser on a WebAssembly-based .NET runtime. And since you can't install .NET into a web browser like you can on a server or developer machine, the project references the Blazor framework using individual package references.

Reviewer 2: ```suggestion
Blazor WebAssembly projects target .NET Standard instead of .NET Core because they run in the browser on a WebAssembly-based .NET runtime. Since you can't install .NET into a web browser like you can on a server or developer machine, the project references the Blazor framework using individual package references.
```

```razor
@page "/counter"
```

Routing in Blazor is handled client-side, not on the server. As the user navigates in the browser, Blazor intercepts the navigation and then renders the component with the matching route. 

The component routes are not currently inferred by the component's file location like they are with *.aspx* pages (although this feature may be added in the future). Each route must be specified explicitly on the component. Putting the routable components in a *Pages* folder has no special meaning and is purely a convention.

We'll look in greater detail at routing in Blazor in the [Pages and routing](./pages-routing.md) section.

## Layout

In ASP.NET Web Forms apps, common page layout is handled using master pages (*Site.Master*). In Blazor apps, page layout is handled using layout components (*Shared/MainLayout.razor*). Layout components will be discussed in more detail in [Page layout](./page-layout) section.

> ## Bootstrapping Blazor

Reviewer 2: ```suggestion
## Bootstrap Blazor
```

@namespace BlazorApp1.Pages
@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>BlazorApp1</title>
    <base href="~/" />
    <link rel="stylesheet" href="css/bootstrap/bootstrap.min.css" />
    <link href="css/site.css" rel="stylesheet" />
</head>
<body>
    <app>
        @(await Html.RenderComponentAsync<App>(RenderMode.ServerPrerendered))
    </app>

    <script src="_framework/blazor.server.js"></script>
</body>
</html>
```

In the Blazor WebAssembly app, the host page is a simple static HTML file under *wwwroot/index.html*. The `<app>` element is used to indicate where the root component should be rendered.

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title>BlazorApp2</title>
    <base href="/" />
    <link href="css/bootstrap/bootstrap.min.css" rel="stylesheet" />
    <link href="css/site.css" rel="stylesheet" />
</head>
<body>
    <app>Loading...</app>

    <script src="_framework/blazor.webassembly.js"></script>
</body>
</html>
```

> The specific component to render is configured in the app's `Startup.Configure` method with a corresponding CSS selector for where the component should be rendered.

Reviewer 2: ```suggestion
The specific component to render is configured in the app's `Startup.Configure` method with a corresponding CSS selector indicating where the component should be rendered.
```

```csharp
    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
        }

        public void Configure(IComponentsApplicationBuilder app)
        {
            app.AddComponent<App>("app");
        }
    }
```

## Build output

When the Blazor projects are built, all of the Razor component files are compiled along with the project code into a single assembly. Unlike ASP.NET Web Forms projects, Blazor doesn't support runtime compilation of the UI logic.

> ## Running the app

Reviewer 2: ```suggestion
## Run the app
```

> To run the Blazor Server app simply hit F5 in Visual Studio. Because Blazor apps do not support runtime compilation, to see the results of code changes, including component markup changes, you typically need to rebuild and restart the app when running under the debugger. If you don't run under the debugger (Ctrl+F5), Visual Studio watches for file changes and restarts the app as changes are made. You simply refresh the browser as changes are made.

Reviewer 2: ```suggestion
To run the Blazor Server app, press `F5` in Visual Studio. Because Blazor apps do not support runtime compilation, to see the results of code changes, including component markup changes, you typically need to rebuild and restart the app when running under the debugger. If you don't run under the debugger (`Ctrl+F5`), Visual Studio watches for file changes and restarts the app as changes are made. You simply refresh the browser as changes are made.
```
